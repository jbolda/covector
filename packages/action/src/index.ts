import * as core from "@actions/core";
import * as github from "@actions/github";
import { covector } from "../../covector/src/run";
import {
  commandText,
  packageListToArray,
  injectPublishFunctions,
  createReleases,
} from "./utils";
import { postGithubComment } from "./comment/postGithubComment";
import fs from "fs";

import type {
  CovectorStatus,
  CovectorVersion,
  CovectorPublish,
} from "../../types/src";
import { formatComment } from "./comment/formatGithubComment";
import type { PullRequestPayload } from "./comment/types";
import type { Operation } from "effection";
import { CommitResponse, getCommitContext } from "./pr/getCommitContext";

export function* run(): Generator<any, any, any> {
  try {
    const cwd =
      core.getInput("cwd") === "" ? process.cwd() : core.getInput("cwd");
    const token =
      core.getInput("token") === ""
        ? process.env.GITHUB_TOKEN || ""
        : core.getInput("token");
    const inputCommand = core.getInput("command");
    const releaseCommitish =
      core.getInput("releaseCommitish") || github.context.sha;

    if (!inputCommand) {
      throw new Error("Must specify command for action. See README.");
    }

    const filterPackages = packageListToArray(core.getInput("filterPackages"));
    let command = inputCommand;

    if (inputCommand === "version-or-publish") {
      const status = yield covector({ command: "status", cwd, logs: false });
      if (status.response === "No changes.") {
        console.log("As there are no changes, let's try publishing.");
        command = "publish";
      } else {
        command = "version";
      }
    }

    core.setOutput("commandRan", command);
    let successfulPublish = false;
    if (command === "status") {
      const covectored: CovectorStatus = yield covector({
        command,
        filterPackages,
        cwd,
      });
      core.setOutput("status", covectored.response);
      core.setOutput("templatePipe", covectored.pipeTemplate);

      core.setOutput(
        `willPublish`,
        covectored.response === "No changes." &&
          "pkgReadyToPublish" in covectored &&
          covectored?.pkgReadyToPublish.length > 0
      );

      if (
        "pkgReadyToPublish" in covectored &&
        covectored?.pkgReadyToPublish?.length > 0
      ) {
        covectored.pkgReadyToPublish.forEach((pkg) => {
          core.setOutput(
            `willPublish-${pkg.pkg}`
              .replace(/\@/g, "-")
              .replace(/\//g, "-")
              .replace(/\_/g, "-"),
            true
          );
          core.setOutput(
            `version-${pkg.pkg}`
              .replace(/\@/g, "-")
              .replace(/\//g, "-")
              .replace(/\_/g, "-"),
            pkg?.pkgFile?.version ?? ""
          );
        });
      }

      if (core.getInput("comment")) {
        if (!token) {
          core.setFailed("The `token` is required to create a comment.");
        } else {
          const octokit = github.getOctokit(token);
          const payload: PullRequestPayload = JSON.parse(
            fs.readFileSync(`${process.env.GITHUB_EVENT_PATH}`, "utf-8")
          );

          if (payload.pull_request) {
            const projectReadmeExists = fs.existsSync(`${cwd}/.changes`);
            const comment = formatComment({
              covectored,
              payload,
              changeFolder: ".changes",
              projectReadmeExists,
            });
            if (comment) yield postGithubComment({ comment, octokit, payload });
          } else {
            console.warn(
              "Comments can only be used on pull requests, skipping."
            );
          }
        }
      }
    } else if (command === "version") {
      const status: CovectorStatus = yield covector({ command: "status", cwd });
      core.setOutput("status", status.response);

      function* createContext({ commits }: { commits: string[] }): Operation<
        Operation<{
          context: Record<string, string>;
          changeContext: Record<string, string>;
        }>
      > {
        let shas = {};
        try {
          const octokit = github.getOctokit(token);
          const prContext: CommitResponse = yield getCommitContext(
            octokit.graphql,
            github.context.repo.owner,
            github.context.repo.repo,
            commits
          );
          shas = Object.entries(prContext.repository).reduce(
            (finalShas, [shaKey, shaContext]) => {
              const reviewers = shaContext.associatedPullRequests.nodes[0]
                ?.reviews?.nodes
                ? shaContext.associatedPullRequests.nodes[0].reviews.nodes.reduce(
                    (reviewersList, reviewer) => {
                      reviewersList[reviewer.author.login] = "APPROVED";
                      return reviewersList;
                    },
                    {} as { [key: string]: string }
                  )
                : {};
              finalShas[shaKey] = {
                author: shaContext.associatedPullRequests.nodes[0].author.login,
                reviewed: Object.keys(reviewers).join(", "),
              };
              return finalShas;
            },
            {} as { [key: string]: { author: string; reviewed: string } }
          );
        } catch (error) {
          // if it fails, continue with context
          console.error(error);
        }
        const context = { ...shas };

        return function* defineContexts() {
          const changeContext = {};
          return { context, changeContext };
        };
      }

      const covectored: CovectorVersion = yield covector({
        command,
        filterPackages,
        cwd,
        ...(core.getInput("recognizeContributors") ? { createContext } : {}),
      });
      core.setOutput("templatePipe", covectored.pipeTemplate);

      const covectoredSmushed = Object.keys(covectored.commandsRan).reduce(
        (text, pkg) => {
          if (typeof covectored.commandsRan[pkg].command === "string") {
            text = `${text}\n\n\n# ${pkg}\n\n${commandText(
              covectored.commandsRan[pkg]
            )}`;
          }
          return text;
        },
        "# Version Updates\n\nMerging this PR will release new versions of the following packages based on your change files.\n\n"
      );
      core.setOutput(
        "change",
        covectoredSmushed.replace(
          /\[`?(.*?)`?\]\(.*?https\:\/\/(www\.)?github\.com.*?\)/gm,
          "$1"
        )
      );
      const payload = JSON.stringify(covectoredSmushed, undefined, 2);
      core.startGroup(`covector version output`);
      console.log(`The covector output: ${payload}`);
      core.endGroup();
    } else if (command === "publish") {
      const status = yield covector({ command: "status", cwd });
      core.setOutput("status", status.response);

      let covectored: CovectorPublish;
      core.debug(
        `createRelease is ${core.getInput("createRelease")} ${
          token ? "with" : "without"
        } a token.`
      );
      if (core.getInput("createRelease") === "true" && token) {
        const octokit = github.getOctokit(token);
        const { owner, repo } = github.context.repo;
        core.debug(`Fetched context, owner is ${owner} and repo is ${repo}.`);
        covectored = yield covector({
          command,
          filterPackages,
          cwd,
          modifyConfig: injectPublishFunctions([
            createReleases({
              core,
              octokit,
              owner,
              repo,
              targetCommitish: releaseCommitish,
            }),
          ]),
        });
      } else {
        covectored = yield covector({
          command,
          filterPackages,
          cwd,
        });
      }

      if (covectored.commandsRan) {
        let packagesPublished = Object.keys(covectored.commandsRan).reduce(
          (pub, pkg) => {
            if (!covectored.commandsRan[pkg].published) {
              return pub;
            } else {
              return pub === "" ? pkg : `${pub},${pkg}`;
            }
          },
          ""
        );
        core.setOutput("packagesPublished", packagesPublished);
        core.setOutput("templatePipe", JSON.stringify(covectored.pipeTemplate));

        for (let pkg of Object.keys(covectored.commandsRan)) {
          if (covectored.commandsRan[pkg].command !== false)
            successfulPublish = true;
        }
        core.setOutput("successfulPublish", successfulPublish);

        core.setOutput("change", covectored.commandsRan);
        const payload = JSON.stringify(covectored.commandsRan, undefined, 2);

        core.startGroup(`covector publish output`);
        console.log(`The covector output: ${payload}`);
        core.endGroup();
        return covectored;
      }
    } else if (command === "preview") {
      const configuredLabel = core.getInput("label");
      const previewLabel =
        github?.context?.payload?.pull_request?.labels?.filter(
          ({ name }: { name: String }) => name === configuredLabel
        ).length;
      const previewVersion = core.getInput("previewVersion");
      const versionIdentifier = core.getInput("identifier");

      if (github.context.eventName !== "pull_request") {
        throw new Error(
          `The 'preview' command for the covector action is only meant to run on pull requests.`
        );
      }

      if (github.context.eventName !== "pull_request") {
        throw new Error(
          `The 'preview' command for the covector action is only meant to run on pull requests.`
        );
      }

      if (!previewLabel) {
        console.log(
          `Not publishing any preview packages because the "${configuredLabel}" label has not been applied to this pull request.`
        );
      } else {
        // primarily runs publish
        let covectored: CovectorPublish;
        const branchName = github?.context?.payload?.pull_request?.head?.ref;
        let identifier;
        let versionTemplate;
        const branchTag = branchName
          .replace(/(?!.\_)\_/g, "__")
          .replace(/\//g, "_");

        if (branchName === "latest") {
          throw new Error(
            `Using the branch name, 'latest', will conflict with restricted tags when publishing packages. Please create another pull request with a different branch name.`
          );
        }

        switch (versionIdentifier) {
          case "branch":
            identifier = branchName.replace(/\_/g, "-").replace(/\//g, "-");
            break;
          default:
            throw new Error(
              `Version identifier you specified, "${versionIdentifier}", is invalid.`
            );
        }

        switch (previewVersion) {
          case "date":
            versionTemplate = `${identifier}.${Date.now()}`;
            break;
          case "sha":
            versionTemplate = `${identifier}.${github.context.payload.after.substring(
              0,
              7
            )}`;
            break;
          default:
            throw new Error(
              `Preview version template you specified, "${previewVersion}", is invalid. Please use 'date' or 'sha'.`
            );
        }

        covectored = yield covector({
          command,
          filterPackages,
          cwd,
          previewVersion: versionTemplate,
          branchTag,
        });

        if (covectored.commandsRan) {
          let packagesPublished: any = Object.entries(
            covectored.commandsRan
          ).reduce((pub: string[], pkg: any[]) => {
            if (pkg[1].published) {
              let { name: pkgName, version: pkgVersion }: any =
                pkg[1].pkg.pkgFile.pkg;
              return pub.concat(`${pkgName}@${pkgVersion}`);
            }
            return pub;
          }, []);

          if (token && github.context.payload.pull_request) {
            const octokit = github.getOctokit(token);
            const {
              pull_request: { number: issue_number },
              repository: {
                name: repo,
                owner: { login: owner },
              },
              after,
            }: any = github.context.payload;

            const { data } = yield octokit.rest.issues.listComments({
              owner,
              repo,
              issue_number,
            });

            const covectorComments = data.filter((comment: any) =>
              comment.body.includes("<!-- covector comment -->")
            );

            let prComment = () => {
              let commentHead = "";
              let commentBody = "";
              if (packagesPublished.length) {
                commentHead = `The following preview packages have been published by Covector:`;
                commentBody = packagesPublished.reduce(
                  (result: string, publishedPackage: string) => {
                    return `${result}- \`${publishedPackage}\`\n`;
                  },
                  ""
                );
              } else {
                commentHead = "Covector did not publish any preview packages.";
              }
              const commentFoot = `<p align="right">${after.slice(0, 7)}</p>`;
              return `${commentHead}\n${commentBody}\n${commentFoot}\n<!-- covector comment -->`;
            };

            if (covectorComments.length !== 1) {
              yield octokit.rest.issues.createComment({
                owner,
                repo,
                issue_number,
                body: prComment(),
              });
            } else {
              yield octokit.rest.issues.updateComment({
                owner,
                repo,
                comment_id: covectorComments[0].id,
                body: prComment(),
              });
            }
          } else {
            throw new Error(
              `Github token argument in preview workflow is missing but it is required in order to generate comments on pull requests.`
            );
          }

          core.setOutput("packagesPublished", packagesPublished);
          core.setOutput("templatePipe", covectored.pipeTemplate);
          for (let pkg of Object.keys(covectored.commandsRan)) {
            if (covectored.commandsRan[pkg].command !== false)
              successfulPublish = true;
          }
          core.setOutput("successfulPublish", successfulPublish);
          core.setOutput("change", covectored.commandsRan);
        }
      }
    } else {
      throw new Error(
        `Command "${command}" not recognized. See README for which commands are available.`
      );
    }
  } catch (error: any) {
    core.setFailed(error.message);
  }
}
