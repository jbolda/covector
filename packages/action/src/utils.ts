import fs from "fs";
import type { FunctionPipe } from "../../types/src";
import type { ConfigFile, Logger } from "@covector/types";
import type { GitHub } from "@actions/github/lib/utils";

export const commandText = (pkg: {
  precommand: string | boolean | null;
  command: string | boolean | null;
  postcommand: string | boolean | null;
}): string => {
  const { precommand, command, postcommand } = pkg;
  let text = "";

  if (typeof precommand !== "boolean") {
    text = `${text}${precommand}\n`;
  }

  if (typeof command !== "boolean") {
    text = `${text}${command}\n`;
  }

  if (typeof postcommand !== "boolean") {
    text = `${text}${postcommand}\n`;
  }

  return text === "" ? "Publish complete." : text;
};

export const packageListToArray = (list: string): string[] => {
  if (list === "" || !list) {
    return [];
  } else {
    return list.split(",");
  }
};

export const injectPublishFunctions = curry(
  (functionsToInject: Function[], config: ConfigFile) => {
    if (!config) return config;
    if (!Array.isArray(functionsToInject))
      throw new Error(
        "injectPublishFunctions() in modifyConfig() expects an array"
      );
    return {
      ...config,
      pkgManagers: injectIntoPublish(config.pkgManagers, functionsToInject),
      packages: injectIntoPublish(config.packages, functionsToInject),
    };
  }
);

const injectIntoPublish = (
  packages: { [k: string]: object } | undefined,
  functionsToInject: Function[]
) => {
  if (!packages) return {};
  return Object.keys(packages).reduce((finalConfig, pkg) => {
    finalConfig![pkg] = Object.keys(packages![pkg]).reduce(
      (pm: { [k: string]: any }, p) => {
        if (p.startsWith("publish") && pm[p]) {
          pm[p] = Array.isArray(pm[p])
            ? pm[p].concat(functionsToInject)
            : [pm[p]].concat(functionsToInject);
        }
        return pm;
      },
      packages![pkg] || {}
    );

    return finalConfig;
  }, packages);
};

function curry(func: Function): Function {
  return function f1(...args1: any[]): Promise<Function> | Function {
    if (args1.length >= func.length) {
      return new Promise((resolve) => resolve(func.apply(null, args1)));
    } else {
      return function f2(...args2: any[]) {
        return f1.apply(null, args1.concat(args2));
      };
    }
  };
}

export const createReleases = curry(
  async (
    {
      logger,
      core,
      octokit,
      owner,
      repo,
      targetCommitish,
    }: {
      logger: Logger;
      core: { [k: string]: Function };
      octokit: InstanceType<typeof GitHub>;
      owner: string;
      repo: string;
      targetCommitish: string;
    },
    pipe: FunctionPipe
  ): Promise<void> => {
    if (!pipe.pkgFile) {
      logger.error(
        `skipping Github Release for ${pipe.pkg}, no package file present`
      );
      return;
    }

    if (!pipe.releaseTag) {
      logger.error(
        `skipping Github Release for ${pipe.pkg}, releaseTag is null`
      );
      return;
    }

    const releaseTag = pipe.releaseTag;
    logger.debug(`creating release with tag ${releaseTag}`);
    const existingRelease = await octokit.rest.repos
      .listReleases({
        owner,
        repo,
      })
      .then((releases) => {
        const release = releases.data.find(
          (r) => r.draft && r.tag_name === releaseTag
        );
        return release ? release : null;
      })
      .catch((error: Error) => null);

    let releaseResponse;
    if (existingRelease && existingRelease.draft) {
      logger.info(
        `updating and publishing Github Release for ${pipe.pkg}@${pipe.pkgFile.version}`
      );
      releaseResponse = await octokit.rest.repos
        .updateRelease({
          owner,
          repo,
          release_id: existingRelease.id,
          body: `${
            existingRelease.body ? `${existingRelease.body}\n` : ""
          }${commandText(pipe.pkgCommandsRan)}`,
          draft: false,
        })
        .then((response) => response.data);
    } else {
      logger.info(
        `creating Github Release for ${pipe.pkg}@${pipe.pkgFile.version}`
      );
      releaseResponse = await octokit.rest.repos
        .createRelease({
          owner,
          repo,
          tag_name: releaseTag,
          name: `${pipe.pkg} v${pipe.pkgFile.version}`,
          body: commandText(pipe.pkgCommandsRan),
          draft: core.getInput("draftRelease") === "true" ? true : false,
          target_commitish: targetCommitish,
        })
        .then((response) => response.data);
    }
    // keeping this one since this was originally used
    // considered deprecated and will remove in v1
    core.setOutput(`${pipe.pkg}-published`, "true");
    // this will be used moving forward
    const cleanPipePkg = pipe.pkg
      .replace(/\@/g, "-")
      .replace(/\//g, "-")
      .replace(/\_/g, "-");
    core.setOutput(`published-${cleanPipePkg}`, "true");

    // output information about the created release
    core.setOutput("releaseUrl", releaseResponse.url);
    core.setOutput("releaseUploadUrl", releaseResponse.upload_url);
    core.setOutput("releaseId", releaseResponse.id);

    core.setOutput(`${cleanPipePkg}-releaseUrl`, releaseResponse.url);
    core.setOutput(
      `${cleanPipePkg}-releaseUploadUrl`,
      releaseResponse.upload_url
    );
    core.setOutput(`${cleanPipePkg}-releaseId`, releaseResponse.id);

    logger.info({
      msg: `github release created for ${pipe.pkg} with id: ${releaseResponse.id}`,
      renderAsYAML: releaseResponse,
    });

    if (pipe.assets) {
      try {
        for (let asset of pipe.assets) {
          logger.info(
            `uploading asset ${asset.name} for ${pipe.pkg}@${pipe.pkgFile.version}`
          );
          const uploadedAsset = await octokit.rest.repos
            .uploadReleaseAsset({
              owner,
              repo,
              release_id: releaseResponse.id,
              name: asset.name,
              // this type seems to be set incorrectly upstream as their API expects a Buffer
              // per https://docs.github.com/en/rest/releases/assets?apiVersion=2022-11-28#upload-a-release-asset
              // or we need to somehow pass in the expected body type but... let's just ignore it
              // @ts-expect-error error TS2322: Type 'Buffer' is not assignable to type 'string'.
              data: fs.readFileSync(asset.path),
            })
            .then((response) => response.data);
          logger.info({
            msg: `asset uploaded to release for ${pipe.pkg}`,
            renderAsYAML: uploadedAsset,
          });
        }
      } catch (error) {
        logger.error(error);
      }
    }
  }
);
