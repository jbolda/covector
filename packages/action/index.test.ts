import * as core from "@actions/core";
import * as github from "@actions/github";
import { run as covector } from "./src";
import { run } from "effection";
import { packageListToArray } from "./src/utils";
import fixtures from "fixturez";
const f = fixtures(__dirname);

jest.mock("@actions/core");
jest.mock("@actions/github", () => ({
  getOctokit: jest.fn(),
  context: { repo: { owner: "genericOwner", repo: "genericRepo" } },
}));

let consoleMock = console as jest.Mocked<Console>;
const mockConsole = (toMock: string[]) => {
  const originalConsole = { ...console };
  debugger;
  toMock.forEach((mock) => {
    (console as any)[mock] = jest.fn();
  });
  consoleMock = console as jest.Mocked<Console>;
  return () => {
    global.console = originalConsole;
  };
};

describe("packageListToArray", () => {
  it("returns empty array on empty string", () => {
    const list = "";
    const pkgArray = packageListToArray(list);
    expect(pkgArray.length).toBe(0);
  });

  it("splits on comma", () => {
    const list = "package1,package2,package3";
    const pkgArray = packageListToArray(list);
    expect(pkgArray[0]).toBe("package1");
    expect(pkgArray[1]).toBe("package2");
    expect(pkgArray[2]).toBe("package3");
  });

  it("considers a single package", () => {
    const list = "package17";
    const pkgArray = packageListToArray(list);
    expect(pkgArray[0]).toBe("package17");
    expect(pkgArray[1]).toBe(undefined);
  });
});

// doing this is an example showing how we can mock the github side effects
describe("test mocks", () => {
  it("mocks core getInput", () => {
    jest
      .spyOn(core, "getInput")
      .mockImplementationOnce((arg) => `This returns ${arg}`);
    const test = core.getInput("test");
    expect(test).toBe("This returns test");
  });

  it("mocks github context", () => {
    jest
      .spyOn(github, "getOctokit")
      //@ts-ignore
      .mockImplementationOnce((arg) => ({
        context: { repo: { owner: "genericOwner", repo: "genericRepo" } },
      }));
    const octokit = github.getOctokit("token");
    expect(octokit).toEqual({
      context: { repo: { owner: "genericOwner", repo: "genericRepo" } },
    });
  });

  it("mocks octokit createRelease", () => {
    jest.spyOn(github, "getOctokit").mockImplementationOnce((arg) => ({
      //@ts-ignore
      repos: { createRelease: (obj) => obj },
    }));
    const octokit = github.getOctokit("token");
    //@ts-ignore
    const releaseResponse = octokit.repos.createRelease({ body: "text" });
    expect(releaseResponse).toEqual({ body: "text" });
  });
});

describe("full e2e test", () => {
  let restoreConsole: Function;
  beforeEach(() => {
    restoreConsole = mockConsole(["log", "dir", "info", "error"]);
    jest.clearAllMocks();
  });
  afterEach(() => {
    restoreConsole();
  });

  jest
    .spyOn(github, "getOctokit")
    // @ts-ignore
    .mockImplementation((token: string) => ({
      context: { repo: { owner: "genericOwner", repo: "genericRepo" } },
      repos: {
        getReleaseByTag: jest.fn((input) => ({ data: input })),
        updateRelease: jest.fn((input) => ({ data: input })),
        createRelease: jest.fn((input) => ({ data: input })),
        uploadReleaseAsset: jest.fn((input) => ({ data: input })),
      },
    }));

  describe("of status", () => {
    it("output", async () => {
      const cwd: string = f.copy("integration.js-with-complex-commands");

      const input: { [k: string]: string } = {
        command: "status",
        cwd: cwd,
        createRelease: "false",
        draftRelease: "false",
        token: "randomsequenceofcharactersforsecurity",
      };

      jest.spyOn(core, "getInput").mockImplementation((arg) => input[arg]);

      const covectoredAction = await run(covector());
      expect(core.setOutput).toHaveBeenCalledWith("commandRan", "status");
      expect(core.setOutput).toHaveBeenCalledWith("status", "No changes.");
    });
  });

  describe("of version", () => {
    it("output", async () => {
      const cwd: string = f.copy("integration.js-with-complex-commands");

      const input: { [k: string]: string } = {
        command: "version",
        cwd: cwd,
        createRelease: "false",
        draftRelease: "false",
        token: "randomsequenceofcharactersforsecurity",
      };

      jest.spyOn(core, "getInput").mockImplementation((arg) => input[arg]);

      const covectoredAction = await run(covector());
      expect(core.setOutput).toHaveBeenCalledWith("status", "No changes.");
      expect(core.setOutput).toHaveBeenCalledWith("commandRan", "version");
      expect(core.setOutput).toHaveBeenCalledWith("successfulPublish", false);
      // to cover template pipe
      expect(core.setOutput).toMatchSnapshot();
    });
  });

  describe("of publish", () => {
    it("input", async () => {
      const cwd: string = f.copy("integration.js-with-complex-commands");

      const input: { [k: string]: string } = {
        command: "publish",
        cwd: cwd,
        createRelease: "true",
        draftRelease: "false",
        token: "randomsequenceofcharactersforsecurity",
      };

      jest.spyOn(core, "getInput").mockImplementation((arg) => input[arg]);

      const covectoredAction = await run(covector());
      expect({
        consoleLog: consoleMock.log.mock.calls,
        consoleDir: consoleMock.dir.mock.calls,
      }).toMatchSnapshot();
      expect(core.setOutput).toMatchSnapshot();
    });

    it("output", async () => {
      const cwd: string = f.copy("integration.js-with-complex-commands");

      const input: { [k: string]: string } = {
        command: "publish",
        cwd: cwd,
        createRelease: "false",
        draftRelease: "false",
        token: "randomsequenceofcharactersforsecurity",
      };

      jest.spyOn(core, "getInput").mockImplementation((arg) => input[arg]);

      const covectoredAction = await run(covector());
      expect(core.setOutput).toHaveBeenCalledWith("status", "No changes.");
      expect(core.setOutput).toHaveBeenCalledWith("commandRan", "publish");
      expect(core.setOutput).toHaveBeenCalledWith("successfulPublish", true);
      expect(core.setOutput).toHaveBeenCalledWith(
        "packagesPublished",
        "package-one,package-two"
      );
      // to cover template pipe
      expect(core.setOutput).toMatchSnapshot();
    });

    it("github release update", async () => {
      const cwd: string = f.copy("integration.js-with-complex-commands");

      const input: { [k: string]: string } = {
        command: "publish",
        cwd: cwd,
        createRelease: "true",
        draftRelease: "false",
        token: "randomsequenceofcharactersforsecurity",
      };

      jest.spyOn(core, "getInput").mockImplementation((arg) => input[arg]);
      const octokit = jest
        .spyOn(github, "getOctokit")
        // @ts-ignore
        .mockImplementation((token: string) => ({
          context: { repo: { owner: "genericOwner", repo: "genericRepo" } },
          repos: {
            getReleaseByTag: jest
              .fn()
              .mockResolvedValueOnce({
                draft: true,
                id: 15,
              })
              .mockResolvedValueOnce({
                draft: true,
                id: 22,
              }) as jest.MockedFunction<any>,
            updateRelease: jest.fn((input) => ({ data: input })),
            createRelease: jest.fn((input) => ({ data: input })),
            uploadReleaseAsset: jest.fn((input) => ({ data: input })),
          },
        }));

      const covectoredAction = await run(covector());
      expect(octokit).toHaveBeenCalledWith(input.token);
      const {
        getReleaseByTag,
        createRelease,
        updateRelease,
        // @ts-ignore
      } = github.getOctokit.mock.results[0].value.repos;
      expect(getReleaseByTag).toHaveBeenCalledWith({
        owner: "genericOwner",
        repo: "genericRepo",
        tag: "package-one-v2.3.1",
      });
      expect(getReleaseByTag).toHaveBeenCalledWith({
        owner: "genericOwner",
        repo: "genericRepo",
        tag: "package-two-v1.9.0",
      });

      expect(updateRelease).toHaveBeenCalledWith({
        owner: "genericOwner",
        repo: "genericRepo",
        release_id: 15,
        draft: false,
        body: "## \\[2.3.1]\n\n- Added some cool things.\n\npublish\n\n",
      });
      expect(updateRelease).toHaveBeenCalledWith({
        owner: "genericOwner",
        repo: "genericRepo",
        release_id: 22,
        draft: false,
        body: "## \\[1.9.0]\n\n- Added some even cooler things.\n\npublish\n\n",
      });

      expect(createRelease).toHaveBeenCalledTimes(0);
    });

    it("github release creation", async () => {
      const cwd: string = f.copy("integration.js-with-complex-commands");

      const input: { [k: string]: string } = {
        command: "publish",
        cwd: cwd,
        createRelease: "true",
        draftRelease: "false",
        token: "randomsequenceofcharactersforsecurity",
      };

      jest.spyOn(core, "getInput").mockImplementation((arg) => input[arg]);
      const octokit = jest
        .spyOn(github, "getOctokit")
        // @ts-ignore
        .mockImplementation((token: string) => ({
          context: { repo: { owner: "genericOwner", repo: "genericRepo" } },
          repos: {
            getReleaseByTag: jest.fn(() => false) as jest.MockedFunction<any>,
            updateRelease: jest.fn((input) => ({ data: input })),
            createRelease: jest.fn((input) => ({ data: input })),
            uploadReleaseAsset: jest.fn((input) => ({ data: input })),
          },
        }));

      const covectoredAction = await run(covector());
      expect(octokit).toHaveBeenCalledWith(input.token);
      const {
        getReleaseByTag,
        createRelease,
        updateRelease,
        // @ts-ignore
      } = github.getOctokit.mock.results[0].value.repos;

      expect(getReleaseByTag).toHaveBeenCalledWith({
        owner: "genericOwner",
        repo: "genericRepo",
        tag: "package-one-v2.3.1",
      });
      expect(getReleaseByTag).toHaveBeenCalledWith({
        owner: "genericOwner",
        repo: "genericRepo",
        tag: "package-two-v1.9.0",
      });

      expect(createRelease).toHaveBeenCalledWith({
        owner: "genericOwner",
        repo: "genericRepo",
        name: "package-one v2.3.1",
        tag_name: "package-one-v2.3.1",
        draft: false,
        body: "## \\[2.3.1]\n\n- Added some cool things.\n\npublish\n\n",
      });
      expect(createRelease).toHaveBeenCalledWith({
        owner: "genericOwner",
        repo: "genericRepo",
        name: "package-two v1.9.0",
        tag_name: "package-two-v1.9.0",
        draft: false,
        body: "## \\[1.9.0]\n\n- Added some even cooler things.\n\npublish\n\n",
      });

      expect(updateRelease).toHaveBeenCalledTimes(0);
    });
  });
});
