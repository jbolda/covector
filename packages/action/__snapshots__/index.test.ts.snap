// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`full e2e test tests publish 1`] = `
Object {
  "consoleDir": Array [],
  "consoleLog": Array [
    Array [
      "package-one [publish]: echo publish",
    ],
    Array [
      "publish",
    ],
    Array [
      "creating Github Release for package-one@2.3.1",
    ],
    Array [
      "release created: ",
      Object {
        "body": "## \\\\[2.3.1]

- Added some cool things.

publish

",
        "draft": false,
        "id": "boop",
        "name": "package-one v2.3.1",
        "owner": "genericOwner",
        "repo": "genericRepo",
        "tag_name": "package-one-v2.3.1",
      },
    ],
    Array [
      "package-two [publish]: echo publish",
    ],
    Array [
      "publish",
    ],
    Array [
      "creating Github Release for package-two@1.9.0",
    ],
    Array [
      "release created: ",
      Object {
        "body": "## \\\\[1.9.0]

- Added some even cooler things.

publish

",
        "draft": false,
        "id": "boop",
        "name": "package-two v1.9.0",
        "owner": "genericOwner",
        "repo": "genericRepo",
        "tag_name": "package-two-v1.9.0",
      },
    ],
    Array [
      "The covector output: {
  \\"package-one\\": {
    \\"precommand\\": false,
    \\"command\\": \\"## \\\\\\\\[2.3.1]\\\\n\\\\n- Added some cool things.\\\\n\\\\npublish\\\\n\\",
    \\"postcommand\\": false,
    \\"pkg\\": {
      \\"pkg\\": \\"package-one\\",
      \\"pkgFile\\": {
        \\"version\\": \\"2.3.1\\",
        \\"versionMajor\\": 2,
        \\"versionMinor\\": 3,
        \\"versionPatch\\": 1,
        \\"pkg\\": {
          \\"name\\": \\"package-one\\",
          \\"version\\": \\"2.3.1\\",
          \\"license\\": \\"MIT\\",
          \\"scripts\\": {
            \\"build\\": \\"npm info tauri@0.8.0 description\\",
            \\"test\\": \\"npm info covector@0.1.0 license\\"
          }
        },
        \\"name\\": \\"package-one\\"
      },
      \\"path\\": \\"./packages/package-one\\",
      \\"manager\\": \\"javascript\\",
      \\"precommand\\": null,
      \\"command\\": [
        {
          \\"command\\": \\"echo publish\\",
          \\"pipe\\": true
        },
        null
      ],
      \\"postcommand\\": null
    },
    \\"published\\": true
  },
  \\"package-two\\": {
    \\"precommand\\": false,
    \\"command\\": \\"## \\\\\\\\[1.9.0]\\\\n\\\\n- Added some even cooler things.\\\\n\\\\npublish\\\\n\\",
    \\"postcommand\\": false,
    \\"pkg\\": {
      \\"pkg\\": \\"package-two\\",
      \\"pkgFile\\": {
        \\"version\\": \\"1.9.0\\",
        \\"versionMajor\\": 1,
        \\"versionMinor\\": 9,
        \\"versionPatch\\": 0,
        \\"pkg\\": {
          \\"name\\": \\"package-two\\",
          \\"version\\": \\"1.9.0\\",
          \\"license\\": \\"MIT\\",
          \\"scripts\\": {
            \\"build\\": \\"echo this command is not piped, it is run from scripts for pk2\\",
            \\"test\\": \\"echo this command is not piped, it is run from the test script\\"
          }
        },
        \\"name\\": \\"package-two\\"
      },
      \\"path\\": \\"./packages/package-two\\",
      \\"manager\\": \\"javascript\\",
      \\"precommand\\": null,
      \\"command\\": [
        {
          \\"command\\": \\"echo publish\\",
          \\"pipe\\": true
        },
        null
      ],
      \\"postcommand\\": null
    },
    \\"published\\": true
  }
}",
    ],
  ],
  "covectoredAction": Object {
    "package-one": Object {
      "command": "## \\\\[2.3.1]

- Added some cool things.

publish
",
      "pkg": Object {
        "command": Array [
          Object {
            "command": "echo publish",
            "dryRunCommand": undefined,
            "pipe": true,
          },
          [Function],
        ],
        "dependencies": undefined,
        "manager": "javascript",
        "path": "./packages/package-one",
        "pkg": "package-one",
        "pkgFile": Object {
          "name": "package-one",
          "pkg": Object {
            "license": "MIT",
            "name": "package-one",
            "scripts": Object {
              "build": "npm info tauri@0.8.0 description",
              "test": "npm info covector@0.1.0 license",
            },
            "version": "2.3.1",
          },
          "version": "2.3.1",
          "versionMajor": 2,
          "versionMinor": 3,
          "versionPatch": 1,
        },
        "postcommand": null,
        "precommand": null,
      },
      "postcommand": false,
      "precommand": false,
      "published": true,
    },
    "package-two": Object {
      "command": "## \\\\[1.9.0]

- Added some even cooler things.

publish
",
      "pkg": Object {
        "command": Array [
          Object {
            "command": "echo publish",
            "dryRunCommand": undefined,
            "pipe": true,
          },
          [Function],
        ],
        "dependencies": undefined,
        "manager": "javascript",
        "path": "./packages/package-two",
        "pkg": "package-two",
        "pkgFile": Object {
          "name": "package-two",
          "pkg": Object {
            "license": "MIT",
            "name": "package-two",
            "scripts": Object {
              "build": "echo this command is not piped, it is run from scripts for pk2",
              "test": "echo this command is not piped, it is run from the test script",
            },
            "version": "1.9.0",
          },
          "version": "1.9.0",
          "versionMajor": 1,
          "versionMinor": 9,
          "versionPatch": 0,
        },
        "postcommand": null,
        "precommand": null,
      },
      "postcommand": false,
      "precommand": false,
      "published": true,
    },
  },
}
`;
